File: frontend/src/api/metrics.ts
/**
 * API client for performance metrics (Step 26: Performance Optimization)
 */

export interface PerformanceMetrics {
  session_start: string;
  cache: {
    [market: string]: {
      hit_rate: number;
      session_hit_rate: number;
      total_requests: number;
      hits: number;
      misses: number;
    };
  };
  ai: {
    total_calls: number;
    total_cost_usd: number;
    avg_cost_per_call: number;
    calls: Array<{
      market: string;
      input_tokens: number;
      output_tokens: number;
      cost_usd: number;
      timestamp: string;
    }>;
  };
  alerts: {
    ai_cost_exceeded: boolean;
    ai_cost_threshold_usd: number;
    total_ai_cost_usd: number;
  };
  targets: {
    cache_hit_rate_target: number;
    ai_cost_per_run_target_usd: number;
    slow_query_threshold_ms: number;
  };
}

export async function getPerformanceMetrics(): Promise<PerformanceMetrics> {
  const response = await fetch('/api/v1/metrics/performance');
  if (!response.ok) {
    throw new Error('Failed to fetch performance metrics');
  }
  return response.json();
}

export async function resetMetrics(): Promise<void> {
  const response = await fetch('/api/v1/metrics/reset', {
    method: 'POST',
  });
  if (!response.ok) {
    throw new Error('Failed to reset metrics');
  }
}

File: backend/app/api/v1/endpoints/metrics.py
"""
Metrics API endpoints for Step 26: Performance Optimization.
Provides real-time performance metrics and alerts.
"""

from fastapi import APIRouter

from app.core.config import get_settings
from app.core.metrics import get_metrics_collector

router = APIRouter(
    prefix="/metrics", tags=["metrics"]
)


@router.get("/performance")
async def get_performance_metrics():
    """
    Get real-time performance metrics.
    Cache hit rates, AI costs, system stats.
    """
    collector = get_metrics_collector()
    summary = collector.get_summary()

    settings = get_settings()
    total_ai_cost = collector.get_total_ai_cost()
    cost_alert = (
        float(total_ai_cost)
        > settings.ai_cost_alert_threshold_usd
    )

    return {
        **summary,
        "alerts": {
            "ai_cost_exceeded": cost_alert,
            "ai_cost_threshold_usd": (
                settings.ai_cost_alert_threshold_usd
            ),
            "total_ai_cost_usd": float(total_ai_cost),
        },
        "targets": {
            "cache_hit_rate_target": 90.0,
            "ai_cost_per_run_target_usd": 0.10,
            "slow_query_threshold_ms": (
                settings.slow_query_threshold_ms
            ),
        }
    }


@router.post("/reset")
async def reset_metrics():
    """Reset in-memory metrics counters."""
    collector = get_metrics_collector()
    collector.reset()
    return {"message": "Metrics reset successfully"}